docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+
+## Reuse
+
+* [Machine Controller Manager](https://github.com/gardener/machine-controller-manager) already has the capability to adopt existing machines in the IaaS layer based on proper machine definitions.
+* [Etcd Backup Restore](https://github.com/gardener/etcd-backup-restore) already backs up the shoot etcd which is running in the shoot namespace of the source seed cluster. We can re-use this completely while restoring the shoot etcd on the destination seed cluster.
+
+## Proposed Solution
+
+### Backup and Restoration
+
+The backup and restoration of the following resources that are currently not stored in the shoot apiserver/etcd is not covered by the [reuse](#reuse) mentioned above.
+
+* Resources from the shoot namespace in the seed.
+  * Terraform resources such as configmaps and secrets.
+  * Machine Controller Manager resources such as machinedeployments, machinesets, machines, machineclasses, secrets etc.
+* Resources from the shoot backup namespace in the seed.
+  * Terraform resources such as configmaps and secrets.
+
+All the additional resources listed above will be backed up in the same object store as the what is used for backing up the shoot etcd.
+
+* The backup will be watch-based and not schedule based.
+* The backup will be full backups at longer intervals followed by incremental backups at smaller intervals similar to the etcd continuous/incremental backup.
+* While taking backup it will take care of dependencies. This is needed since while restoring from it, validation checks needs to be passed.
+    
+    * But the granularity of the changes stored at the incremental backups would be at the object level and not the actual field-level changes. This is different from the etcd continuous/incremental backups where the incremental changes are recorded at the field level.
+
+#### Components of the solution
+
+Taking the idea from the framework setup for etcd-backup-restore, we will introduce new component named 
+**shoot-backup-restore** per shoot in seed namespace. This will backup the control plane component of shoot only from seed namespace.
+
+##### shoot-backup-restore implementation
+
+* It will get list of objects to be backed up via configmap shoot-resources.
+
+  * TODO: Come up json format for dependencies configmap
+  * Initial thought (not full proof) is To have nested dependency fields list  
+  * Example configmap content:
+  ```yaml
+  awsmachineclass: spec.sercretRef
+  machines: spec.class
+  machineset: ...
+  machinedeployment: ...
+  configmap: ...
+  secret: ...
+  roles: ...       # May not be required?
+  rolebinding: ... # May not be Required?
+  service: ...     # May not be Required?
+
+  ```
+
+  * TODO: Give thought about OwnerReferences
+  * TODO: Think on dependency hooks
+
+* It will apply watch on seed's apiserver for those objects and takes backup of full object including its dependency.
+
+* It will have an `/initialize` endpoint to initialize cluster.
```

Comment:
How this thing can work? It must be a public endpoint?

Reply:
Rather than public i would rephrase it, It need's to be accessible by gardener. 

Said that, I have 60-40%  opinion on where should this deployment run,
-  seed namespace (on seed cluster)
    - Pros: backup operation will continue even if garden cluster is down.
    - Cons: We will again need some synchronisation between pod running in old and new seed cluster(which could be handled by other major part of this PR).
-  project namespace (on garden cluster)
   - Pros: Independent of seed, in case seed change, it just has to recreate seed client.
   - Cons:  Gardener cluster need to be HA.